database/rrdhost.c

```diff
@@ -784,8 +784,11 @@ void rrdhost_free(RRDHOST *host) {
                 netdata_thread_cancel(host->receiver->thread);
             netdata_mutex_unlock(&host->receiver_lock);
             struct receiver_state *rpt = host->receiver;
-            while (host->receiver && !rpt->exited)
+            int i;
+            for (i = 0 ; host->receiver && !rpt->exited && i < 20 ; ++i)
                 sleep_usec(50 * USEC_PER_MS);
+            if (20 == i)
+                info("Deadlock resolved during shutdown of streaming receiver context for host %s.", host->hostname);
```

Comment:
why not make the thread joinable and wait for it that way? Then destroying receiver state

Reply:
The problem is that `join()` will block for ever if we hit this specific race condition, because the thread will never be canceled. Fixing the race condition will probably take a lot of refactoring of the agent, it's not obvious to me.